### PR TITLE
[C++] Append column

### PR DESCRIPTION
**Pull request by eddyxu**
Minimal support of appending a column in C++ `LanceDataset`

### CODE REVIEW

File: cpp/include/lance/arrow/updater.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  ///
+  /// The user must consume the returned results, by calling `Update`, before calling `Next()`
+  /// again.
+  ///
+  /// \return `::arrow::RecordBatch` if success.
+  ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Next();
+
+  /// Update the values to new values, presented in the array.
+  /// The array must has the same length as the batch returned previously via `Next()`.
+  ::arrow::Status Update(const std::shared_ptr<::arrow::Array>& arr);
```
Comment (changhiskhan) [inline cpp/include/lance/arrow/updater.h:66]: maybe call this UpdateBatch to make it clear the input is supposed to line up with batch?

File: cpp/src/lance/arrow/fragment.cc [hunk 2]
```diff
@@ -101,6 +111,10 @@ ::arrow::Result<std::vector<LanceFragment::FileReaderWithSchema>> LanceFragment:
 
 const std::shared_ptr<format::Schema>& LanceFragment::schema() const { return manifest_->schema(); }
 
+const std::shared_ptr<lance::format::DataFragment>& LanceFragment::data_fragment() const {
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/fragment.cc:114]: i keep forgetting what is DataFragment vs LanceFragment

File: cpp/src/lance/io/reader.cc [hunk 3]
```diff
@@ -141,7 +142,7 @@ Status FileReader::Open() {
   assert(metadata_->page_table_position() >= size - kPrefetchSize);
 
   auto num_batches = metadata_->num_batches();
-  auto num_columns = manifest_->schema()->GetFieldsCount();
+  auto num_columns = ranges::max(manifest_->schema()->GetFieldIds()) + 1;
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:145]: don't have to take care of it now but how would this work for column deletions?

File: cpp/src/lance/arrow/updater.cc [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  const auto& fs() const { return dataset_->impl_->fs; }
+
+  /// Prepare to read the next fragment.
+  ::arrow::Status NextFragment();
+
+  std::shared_ptr<LanceDataset> dataset_;
+  /// The schema of the newly generated dataset.
+  std::shared_ptr<lance::format::Schema> full_schema_;
+  /// The schema of the column to updated/added
+  std::shared_ptr<lance::format::Schema> column_schema_;
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/updater.cc:77]: so this can support adding multiple columns at a time?

File: cpp/src/lance/arrow/updater.cc [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+  }
+  while (!last_batch_ && fragment_it_ != fragments_.end()) {
+    if (!writer_) {
+      ARROW_RETURN_NOT_OK(NextFragment());
+    }
+
+    auto fut = batch_generator_();
+    ARROW_ASSIGN_OR_RAISE(last_batch_, fut.result());
+    if (!last_batch_) {
+      ARROW_RETURN_NOT_OK(writer_->Finish().result());
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/updater.cc:134]: i thought this was reading batches from original dataset. why is a writer involved here?

**Review by changhiskhan — APPROVED** [review]
minor questions

File: cpp/include/lance/arrow/updater.h (hunk 1 above)
Comment (eddyxu) [reply cpp/include/lance/arrow/updater.h:66]: sgtm.

File: cpp/src/lance/arrow/fragment.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/fragment.cc:114]: `DataFragment` is the POD, which maps to what is serialized.

`LanceFragment` is subclass of `arrow::Fragment`, which offers the I/O functions, i.e., `ScanBatchAsync`, `CountRows` and etc.

File: cpp/src/lance/io/reader.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:145]: yea, so for now, we dont reclaim the deleted column Ids, so they are still exist in the lookup table, but do not exist in the schema. 

A simplest solution could be just run a new snapshot of everything and reclaim the column IDs.

File: cpp/src/lance/arrow/updater.cc (hunk 4 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/updater.cc:77]: In theory, yes. 

This is internal API, we could change it to be `shared_ptr<lance::Field>` if we'd live to be more strict for now.

File: cpp/src/lance/arrow/updater.cc (hunk 5 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/updater.cc:134]: An updater is basically an iterator to hold a "Batch Generator" and "Batch Writer", the two of which will be opened / closed together to make data aligned between "Data Files".

**Pull request merged**